static/js/drafts.js

```diff
@@ -0,0 +1,227 @@
+
+var exports = {};
+
+var draft_model = (function () {
+    var exports = {};
+
+    // the key that the drafts are stored under.
+    var KEY = "drafts";
+    var ls = localstorage();
+    ls.version = 1;
+
+    function getTimestamp() {
+        return new Date().getTime();
+    }
+
+    function get() {
+        return ls.get(KEY) || {};
+    }
+    exports.get = get;
+
+    exports.getDraft = function (id) {
+        return get()[id] || false;
+    };
+
+    function save(drafts) {
+        ls.set(KEY, drafts);
+    }
+
+    exports.addDraft = function (draft) {
+        var drafts = get();
+
+        // use the base16 of the current time + a random string to reduce
+        // collisions to essentially zero.
+        var id = getTimestamp().toString(16) + "-" + Math.random().toString(16).split(/\./).pop();
+
+        draft.updatedAt = getTimestamp();
+        drafts[id] = draft;
+        save(drafts);
+
+        return id;
+    };
+
+    exports.editDraft = function (id, draft) {
+        var drafts = get();
+
+        if (drafts[id]) {
+            draft.updatedAt = getTimestamp();
+            drafts[id] = draft;
+            save(drafts);
+        }
+    };
+
+    exports.deleteDraft = function (id) {
+        var drafts = get();
+
+        delete drafts[id];
+        save(drafts);
+    };
+
+    return exports;
+}());
+
+exports.draft_model = draft_model;
+
+exports.update_draft = function () {
+    var draft = compose.snapshot_message();
+    var draft_id = $("#new_message_content").data("draft-id");
+
+    if (draft_id !== undefined) {
+        if (draft !== undefined) {
+            draft_model.editDraft(draft_id, draft);
+        } else {
+            draft_model.deleteDraft(draft_id);
+        }
+    } else {
+        if (draft !== undefined) {
+            var new_draft_id = draft_model.addDraft(draft);
+            $("#new_message_content").data("draft-id", new_draft_id);
+        }
+    }
+};
+
+exports.delete_draft_after_send = function () {
+    var draft_id = $("#new_message_content").data("draft-id");
+    if (draft_id) {
+        draft_model.deleteDraft(draft_id);
+    }
+    $("#new_message_content").removeData("draft-id");
+};
+
+exports.restore_draft = function (draft_id) {
+    var draft = draft_model.getDraft(draft_id);
+    if (!draft) {
+        return;
+    }
+
+    var draft_copy = _.extend({}, draft);
+    if ((draft_copy.type === "stream" &&
+         draft_copy.stream.length > 0 &&
+             draft_copy.subject.length > 0) ||
+                 (draft_copy.type === "private" &&
+                  draft_copy.reply_to.length > 0)) {
+        draft_copy = _.extend({replying_to_message: draft_copy},
+                              draft_copy);
+    }
+
+    exports.close();
+    compose_fade.clear_compose();
+    if (draft.type === "stream" && draft.stream === "") {
+      draft_copy.subject = "";
+    }
+    compose.start(draft_copy.type, draft_copy);
+    $("#new_message_content").data("draft-id", draft_id);
```

Comment:
I think you need to do a compose autosize call here.  Otherwise multi-line messages restore like this:

![image](https://cloud.githubusercontent.com/assets/2746074/23194633/1f246b84-f864-11e6-9f46-7efc43814682.png)